airbyte-cdk/python/airbyte_cdk/connector_builder/message_grouper.py

```diff
@@ -45,6 +45,18 @@ def __init__(self, max_pages_per_slice: int, max_slices: int, max_record_limit:
         self._max_slices = max_slices
         self._max_record_limit = max_record_limit
 
+    def _to_nested_and_composite_field(self, field: Optional[Union[str, List[str], List[List[str]]]]) -> List[List[str]]:
+        if not field:
+            return [[]]
+
+        if isinstance(field, str):
+            return [[field]]
+
+        if isinstance(field[0], str):
+            return [field]  # type: ignore  # the type of field is expected to be List[str] here
```

Comment:
You are right for the primary key, thanks for pointing that out. I've always hated the typing of this field especially since [it means something else for the cursor_field](https://github.com/airbytehq/airbyte/blob/44f784e200fd66d43e0f948858aa2a2ee184fef7/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py#L231-L237). The fact that even with typing and MyPy, we can't distinguish between both case (`List[str]` being a nested field or a composite one). At some point, we should have dedicated classes to represent fields. [There was an attempt to do that in the integration tests framework](https://github.com/airbytehq/airbyte/blob/44f784e200fd66d43e0f948858aa2a2ee184fef7/airbyte-cdk/python/airbyte_cdk/test/mock_http/response_builder.py#L30-L74).

All that to say: I've fixed the issue by having two methods: `_pk_to_nested_and_composite_field` and `_cursor_field_to_nested_and_composite_field`.